packages/cursorless-vscode/src/CommandHistory.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  ActionDescriptor,
+  CommandComplete,
+  CommandHistoryItem,
+  Disposable,
+  FileSystem,
+  ReadOnlyHatMap,
+} from "@cursorless/common";
+import type {
+  CommandRunner,
+  CommandRunnerDecorator,
+} from "@cursorless/cursorless-engine";
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as vscode from "vscode";
+
+const filePrefix = "cursorlessCommandHistory";
+const settingSection = "cursorless";
+const settingName = "commandHistory";
+const settingFullName = `${settingSection}.${settingName}`;
+
+export class CommandHistory implements CommandRunnerDecorator {
+  private readonly dirPath: string;
+  private readonly cursorlessVersion: string;
+  private disposable: Disposable;
+  private isActive: boolean = false;
+
+  constructor(
+    extensionContext: vscode.ExtensionContext,
+    fileSystem: FileSystem,
+  ) {
+    this.cursorlessVersion = extensionContext.extension.packageJSON.version;
+    this.dirPath = fileSystem.cursorlessCommandHistoryDirPath;
+
+    // Read initial setting value. The watcher below will take care of changes.
+    this.evaluateSetting();
+
+    this.disposable = vscode.workspace.onDidChangeConfiguration((event) => {
+      if (event.affectsConfiguration(settingFullName)) {
+        this.evaluateSetting();
+      }
+    });
+  }
+
+  wrapCommandRunner(
+    readableHatMap: ReadOnlyHatMap,
+    runner: CommandRunner,
+  ): CommandRunner {
+    if (!this.isActive) {
+      return runner;
+    }
+
+    return {
+      run: async (commandComplete: CommandComplete) => {
+        await this.appendToLog(commandComplete);
+
+        return await runner.run(commandComplete);
+      },
+    };
+  }
+
+  private async appendToLog(command: CommandComplete): Promise<void> {
+    const date = new Date();
+    const fileName = `${filePrefix}_${getMonthDate(date)}.jsonl`;
+    const file = path.join(this.dirPath, fileName);
+
+    const historyItem: CommandHistoryItem = {
+      date: getDayDate(date),
+      cursorlessVersion: this.cursorlessVersion,
+      command: sanitizeCommand(command),
+    };
+    const data = JSON.stringify(historyItem) + "\n";
+
+    await fs.mkdir(this.dirPath, { recursive: true });
+    await fs.appendFile(file, data, "utf8");
```

Comment:
Does this open file in exclusive mode? If not, would be worth doing so

Reply:
@josharian what did you mean when you said "open the file in exclusive mode"? From what I understand, exclusive just means fail if the file already exists, which I don't think is what we want here

Reply:
sorry about the confusion. I always mess this up. ideally we'd use something like a lockfile to coordinate access, e.g. https://www.npmjs.com/package/proper-lockfile (found via google search, no personal experience). it's either that or hope, which isn't necessarily a terrible approach here.

Reply:
Yeah honestly just "hope" might be fine, since I'm not sure I see how you'd end up with multiple windows writing at once

Reply:
yeah. talon and the command file are both serving as de facto external locking mechanisms here. if we end up with corrupt logs, we can reconsider.

Reply:
And even by keyboard, you won't be typing into multiple windows

Reply:
one certainly hopes. people do weird things. (and win weird prizes when they do.)

Reply:
Then I guess we're happy?

Reply:
Yeah I think let's leave it and keep an eye out for corrupt json. Not seeing any better solution that's worth the effort